### DISCUSSION TITLE
Memory References not opaque?

### DISCUSSION THREAD

**Discussion by mblout** [microsoft/vscode-discussions#712] [Extension Development QnA] [answered]
A combination DAP/vscode question...
I'm developing a debug extension, and currently looking at implementing the disassembleRequest in my debug adapter.
Our disassembly is a different than typical; each 'thread' should display a completely different disassembly - its is not one large address space of disassembly, where each "thread" just focuses the proper instruction.

the Debug Adapter Protocol API doc is a little vague, but from this [discussion](https://github.com/microsoft/debug-adapter-protocol/issues/42) I inferred that "memory references" were treated as opaque strings?

The UI gesture for "Open Disassembly View" is on the Call Stack context menu, and passes the "instructionPointerReference" as the "memoryReference" to the adapter disassembleRequest.  There is no "thread" context passed, other than that reference.

So, I was hoping to encode the thread info into the "instructionPointerReference" as a parse-able string, with non-numeric characters, but this doesn't work

DisassembleView.getIndexFromAddress tries to create a BigInt from the instructionAddress, which causes an exception if it is not parse-able as a number.

Q: are memory addresses meant to be opaque, and is this a bug?
Q: if not opaque, should the DAP documentation be more clear?
Q: shouldn't the DAP API for disassembleRequest include an optional 'thread' param (and should vscode use it)?

**Comment by isidorn** [comment]
I am not sure. But here's the code pointer for the DissasemblyView and I think that should help 
https://github.com/microsoft/vscode/blob/fceb5c1f72ab8a70d302d7bf9f00dcf96f4b4e21/src/vs/workbench/contrib/debug/browser/disassemblyView.ts#L47

If you still have questions do let me know and I can ask around. Thank you

**Reply by olguzzar** [reply]
Hi @mblout, did the above code help? If not, please provide additional information so we can further help - thanks!

**Reply by mblout** [reply]
Thank you both for your replies; no, i was already deep in the above code when I asked the original questions. 
My questions were reallt to try to get clarity on the _intent_, but from what I can see, I would state:
- Debug Adapter Protocol spec intends that all memory references _should_ be opaque _strings_.
- VS Code is trying to parse it as a number, which means it does _not_ treat is as opaque; it expects it to be a numeric string, and compares them. If the adapter implementation puts a non-numeric string, the parse fails, and the comparison fails, and features fail (in my case , the disassembly view).
So, having dug into the code, it is not agreeing with the spec. And i was hoping to have someone confirm my understanding of the spec, before i logged an issue.

I'm looking at other areas right now, so I don't have the specific details in my head, but I'd be happy to try to answer specific questions.
Thanks again!

**Reply by burkeholland** [reply]
Pinging @dbaeumer

**Reply by connor4312** [reply, marked as answer]
>Q: are memory addresses meant to be opaque, and is this a bug?

Yes, this is correct. Some languages, like JavaScript, have binary types that can be viewed, but don't expose memory offsets or ways to read arbitrary memory offsets, so having them be a memory address is too restrictive.

> Q: if not opaque, should the DAP documentation be more clear?

We could clarify this.

> VS Code is trying to parse it as a number, which means it does not treat is as opaque; it expects it to be a numeric string, and compares them. If the adapter implementation puts a non-numeric string, the parse fails, and the comparison fails, and features fail (in my case , the disassembly view).

I think is is a misbehavior in VS Code. memoryReferences in the context of read/write memory _are_ treated as opaque. And that's why we specify the formatting of the returned/disassembled `address`es but not memoryReferences. Opened an issue for it: https://github.com/microsoft/vscode/issues/191925
